### PR TITLE
Preserve controlled inner blocks on RESET_BLOCKS

### PR DESCRIPTION
**Pull request by jsnajdr**
Experiment inspired by #61480. `RESET_BLOCKS` action destroys the `controlledInnerBlocks` state and also doesn't carry over the info about controlled inner blocks. `useBlockSync` then needs special code to quickly recover the controlled state and to set the controller inner blocks again.

This PR attempts to improve the `RESET_BLOCKS` action so that the controlled blocks state is carried over. That means:
- `state.controlledInnerBlocks` is preserved, and we continue to know that a block with certain `clientId` should be controlled.
- the state (`state.attributes`, `state.byClientId`, ...) for controlled inner blocks is preserved. Previously only the state for the new blocks and their uncontrolled children (`innerBlocks`) was part of the new state, the rest was forgotten.
- if block has controlled inner blocks, its `state.order` contains the controlled blocks, it's not reset to uncontrolled. These are still in `state.tree`, and will be written to `state.order` if the block switches to uncontrolled mode.

Let's see how this fares with the e2e tests.

Currently the `RESET_BLOCKS` reducer carries over too much info: it keeps everything that was in the previous state, and only adds info from `action.blocks`. This is harmless, but inefficient. If the experiment proves viable, I'll improve the code to carry over only the old controlled blocks + new blocks from `action.blocks`.

### CODE REVIEW

**Comment by youknowriad** [comment]
> state.controlledInnerBlocks is preserved, and we continue to know that a block with certain clientId should be controlled.

What guarantees that the clientIDs of the new blocks are still "controlled"?

**Comment by jsnajdr** [comment]
> What guarantees that the clientIDs of the new blocks are still "controlled"?

First I assume that the `clientId` uniquely identifies a certain block. `RESET_BLOCKS` with a `clientId` that already exists in the tree is assumed to be a hard reset to a different version of the same block, not a replacement with a completely different block, possibly of a different type.

If the new version of the block doesn't want to be controlled, it will signal that by _not_ mounting `useBlockSync` when it renders. Then the old `useBlockSync` that is currently mounted will unmount, and in its cleanup effects it will reset the block to uncontrolled. If `RESET_BLOCKS` specified a block with explicit uncontrolled `innerBlocks`, these will become active after `useBlockSync` resets the `controlledInnerBlocks` state for it.

In short, the controlled state is now controlled solely by `useBlockSync`. On mount, it will set the `controlledInnerBlocks` state and set the controlled inner blocks. During its lifetime it will sync the controlled blocks. And on unmount it will reset the `controlledInnerBlocks` state, which makes the old uncontrolled inner blocks active again.

Until now not everything was under `useBlockSync` control. There was the `RESET_BLOCKS` action that could clear the `controlledInnerBlocks` state on its own.

**Comment by youknowriad** [comment]
>In short, the controlled state is now controlled solely by useBlockSync. On mount, it will set the controlledInnerBlocks state and set the controlled inner blocks. During its lifetime it will sync the controlled blocks. And on unmount it will reset the controlledInnerBlocks state, which makes the old uncontrolled inner blocks active again.
> Until now not everything was under useBlockSync control. There was the RESET_BLOCKS action that could clear the controlledInnerBlocks state on its own.

That makes sense, in my understanding this PR was removing the "cleanup" from useBlockSync while useBlockSync is the only place that knows for sure whether a block is controlled or not.

As for the clientId, for me we can't really assume that a clientId refers to the same block after you call RESET_BLOCKS but this is fine as long as useBlockSync is responsible for cleaning things up.

**Comment by jsnajdr** [comment]
> As for the clientId, for me we can't really assume that a clientId refers to the same block after you call RESET_BLOCKS but this is fine

Yes: even if `RESET_BLOCKS` replaces the `clientId` with an arbitrary block, the course of events is:
- after `RESET_BLOCKS`, `state.controlledInnerBlocks[ clientId ]` is still `true`, `state.tree[ 'controlled||' + clientId ]` contains the old controlled inner blocks, and `state.tree[ clientId ]` contains the new uncontrolled inner blocks (but they are overshadowed).
- at this moment, `getBlocks( clientId )` would return the _old_ controlled blocks, although the new block might be incompatible with that. **I'm not 100% sure if this is fine, but hopefully it is.**
- if the new block has a different type, or doesn't want to be controlled, the old `useBlockSync` will start unmounting. Only at this moment the `state.controlledInnerBlocks[ clientId ]` will be set to `false`, the `controlled||` state will be removed.
- at this moment, after React unmounts the old edit UI, `getBlocks( clientId )` finally returns the right result.

The old implementation didn't have this inconsistent "new block with old controlled children" state. Instead, if both the old and new block were of the same type and both wanted to be controlled (canonical example of such a reset is an undo that changes only other blocks), the block became briefly uncontrolled, with uncontrolled inner blocks. `useBlockSync` had to recover the controlled state in a React effect.

**Comment by Mamaduka** [comment]
Just noting that #75458 was merged into trunk.

**Comment by youknowriad** [comment]
Let's close this, if there's anything missing from what landed in @jsnajdr let us know.
